snapcraft/models/project.py

```diff
@@ -436,13 +437,36 @@ def _validate_autostart_name(cls, name):
 
         return name
 
-    @pydantic.field_validator("bus_name")
+    @pydantic.field_validator(
+        "command", "stop_command", "post_stop_command", "reload_command", "bus_name"
+    )
     @classmethod
-    def _validate_bus_name(cls, name):
-        if not re.match(r"^[A-Za-z0-9/. _#:$-]*$", name):
-            raise ValueError(f"{name!r} is not a valid bus name")
+    def _validate_command(cls, command: str, info: pydantic.ValidationInfo) -> str:
+        # Find any invalid characters in the command.
+        # The regex below is derived from snapd's validator code, modified to be the inverse (^).
+        # https://github.com/canonical/snapd/blob/0706e2d0b20ae2bf030863f142b8491b66e80bcb/snap/validate.go#L756
+        bad_chars: set[str] = set()
+        for bad_char in re.finditer(r"[^A-Za-z0-9/. _#:$-]", command):
+            bad_chars.add(bad_char.group(0))
+
+        if len(bad_chars) > 0:
+            # Guaranteed not-none as the pydantic field_validator decorator always populates it.
+            deserialized = cast(str, info.field_name).replace("_", "-")
+            bad_chars_print = ", ".join(sorted(bad_chars))
+            doc_link = (
+                "https://snapcraft.io/docs/snapcraft-yaml-schema#p-21225-appsapp-name"
+            )
+            message = textwrap.dedent(
+                f"""\
+                {deserialized}: App commands must consist only of alphanumeric characters, spaces, and the following characters:
+                / . _ # : $ -
+                The following characters were encountered, but not allowed: {bad_chars_print}
+                For more complete details about the `command` key, see: {doc_link}
+            """
+            )
+            raise ValueError(message)
```

Comment:
I'm not sure how `snapcraft` is doing this, but `pydantic` in general gathers all the `ValueErrors` and puts then into a separate bullet points for every field that is incorrect. If that's the case in here it may create a weird end-result

Do you have any example of the `cli` run, how this is displayed to the end user ?

Reply:
With these inside of a basic app:
```
command: bird --chirps=5
daemon: "simple"
stop-command: bird --chirps=1337 --name="87U3J@Y"
```

I get the following (omitting the preceding traceback, which would be addressed by https://github.com/canonical/craft-application/pull/559 eventually)
```
pydantic_core._pydantic_core.ValidationError: 2 validation errors for Project
apps.myapp.command
  Value error, command: App commands must consist only of alphanumeric characters, spaces, and the following characters:
/ . _ # : $ -
The following characters were encountered, but not allowed: =
For more complete details about the `command` key, see: https://snapcraft.io/docs/snapcraft-yaml-schema#p-21225-appsapp-name
 [type=value_error, input_value='bird --chirps=5', input_type=str]
    For further information visit https://errors.pydantic.dev/2.9/v/value_error
apps.ansel.stop-command
  Value error, stop-command: App commands must consist only of alphanumeric characters, spaces, and the following characters:
/ . _ # : $ -
The following characters were encountered, but not allowed: ", =, @
For more complete details about the `command` key, see: https://snapcraft.io/docs/snapcraft-yaml-schema#p-21225-appsapp-name
 [type=value_error, input_value='bird --chirps=1337 --name="87U3J@Y"', input_type=str]
    For further information visit https://errors.pydantic.dev/2.9/v/value_error
```

Reply:
Interesting.  I like the level of detail about what is wrong, but agree with @dariuszd21 that multi-line errors aren't presented well by pydantic.  May be better to have a terse message and allow the user to figure out the offending characters on their own.